File: lib/src/utils/convert_utils.dart
import 'package:photo_manager/photo_manager.dart';

class ConvertUtils {
  static List<AssetPathEntity> convertPath(
    Map data, {
    int type = 0,
    DateTime dt,
    FilterOptionGroup optionGroup,
  }) {
    List<AssetPathEntity> result = [];

    List list = data["data"];

    for (final Map item in list) {
      final entity = AssetPathEntity(filterOption: optionGroup)
        ..id = item["id"]
        ..name = item["name"]
        ..typeInt = type
        ..isAll = item["isAll"]
        ..fetchDatetime = dt
        ..assetCount = item["length"];

      result.add(entity);
    }

    return result;
  }

  static List<AssetEntity> convertToAssetList(Map data) {
    List<AssetEntity> result = [];

    List list = data["data"];
    for (final Map item in list) {
      final entity = AssetEntity()
        ..id = item["id"]
        ..path = item["path"]
        ..createDtSecond = item["createDt"]
        ..width = item["width"]
        ..height = item["height"]
        ..duration = item["duration"]
        ..modifiedDateSecond = item["modifiedDt"]
        ..typeInt = item["type"]
        ..longitude = item["lng"]
        ..latitude = item["lat"]
        ..title = item["title"]
        ..fileSize = item["fileSize"];

      result.add(entity);
    }

    return result;
  }

  static AssetEntity convertToAsset(Map map) {
    if (map == null) {
      return null;
    }

    Map data = map["data"];

    final entity = AssetEntity()
      ..id = data["id"]
      ..path = data["path"]
      ..createDtSecond = data["createDt"]
      ..width = data["width"]
      ..height = data["height"]
      ..duration = data["duration"]
      ..modifiedDateSecond = data["modifiedDt"]
      ..typeInt = data["type"]
      ..longitude = data["lng"]
      ..latitude = data["lat"]
      ..title = data["title"]
      ..fileSize = data["fileSize"];

    return entity;
  }
}

File: android/src/main/kotlin/top/kikt/imagescanner/core/utils/ConvertUtils.kt
package top.kikt.imagescanner.core.utils

import top.kikt.imagescanner.AssetType
import top.kikt.imagescanner.core.entity.*

/// create 2019-09-05 by cai


object ConvertUtils {
  fun convertToGalleryResult(list: List<GalleryEntity>): Map<String, Any> {
    val data = ArrayList<Map<String, Any>>()

    for (entity in list) {
      val element = mapOf(
              "id" to entity.id,
              "name" to entity.name,
              "length" to entity.length,
              "isAll" to entity.isAll
      )

      if (entity.length > 0) {
        data.add(element)
      }
    }

    return mapOf(
            "data" to data
    )
  }

  fun convertToAssetResult(list: List<AssetEntity>): Map<String, Any?> {
    val data = ArrayList<Map<String, Any?>>()

    for (entity in list) {
      val element = mapOf(
          "id" to entity.id,
          "path" to entity.path,
          "duration" to entity.duration / 1000,
          "type" to entity.type,
          "createDt" to entity.createDt / 1000,
          "width" to entity.width,
          "height" to entity.height,
          "modifiedDt" to entity.modifiedDate,
          "lat" to entity.lat,
          "lng" to entity.lng,
          "title" to entity.displayName,
          "fileSize" to entity.fileSize
      )
      data.add(element)
    }

    return mapOf(
            "data" to data
    )
  }

  fun convertToAssetResult(entity: AssetEntity): Map<String, Any?> {

    val data = mapOf(
        "id" to entity.id,
            "path" to entity.path,
        "duration" to entity.duration,
        "type" to entity.type,
        "createDt" to entity.createDt / 1000,
        "width" to entity.width,
        "height" to entity.height,
        "modifiedDt" to entity.modifiedDate,
        "lat" to entity.lat,
        "lng" to entity.lng,
        "title" to entity.displayName,
        "fileSize" to entity.fileSize
    )

    return mapOf(
            "data" to data
    )
  }

  private fun getOptionWithKey(map: Map<*, *>, key: String): FilterCond {
    if (map.containsKey(key)) {
      val value = map[key]
      if (value is Map<*, *>) {
        return convertToOption(value)
      }
    }
    return FilterCond()
  }

  fun getOptionFromType(map: Map<*, *>, type: AssetType): FilterCond {
    if (type == AssetType.Video) {
      return getOptionWithKey(map, "video")
    } else if (type == AssetType.Image) {
      return getOptionWithKey(map, "image")
    }
    return FilterCond()
  }

  private fun convertToOption(map: Map<*, *>): FilterCond {
    val filterOptions = FilterCond()
    filterOptions.isShowTitle = map["title"] as Boolean

    val sizeConstraint = FilterCond.SizeConstraint()
    filterOptions.sizeConstraint = sizeConstraint
    val sizeMap = map["size"] as Map<*, *>
    sizeConstraint.minWidth = sizeMap["minWidth"] as Int
    sizeConstraint.maxWidth = sizeMap["maxWidth"] as Int
    sizeConstraint.minHeight = sizeMap["minHeight"] as Int
    sizeConstraint.maxHeight = sizeMap["maxHeight"] as Int

    val durationConstraint = FilterCond.DurationConstraint()
    filterOptions.durationConstraint = durationConstraint
    val durationMap = map["duration"] as Map<*, *>
    durationConstraint.min = (durationMap["min"] as Int).toLong()
    durationConstraint.max = (durationMap["max"] as Int).toLong()

    return filterOptions
  }


  fun convertFilterOptionsFromMap(map: Map<*, *>): FilterOption {
    return FilterOption(map)
  }
}